deriva/core/deriva_binding.py

```diff
@@ -1,12 +1,82 @@
+
+import uuid
+import sys
+import os
+import json
 import requests
 from multiprocessing import Queue
-from . import get_new_requests_session, ConcurrentUpdate, NotModified, DEFAULT_HEADERS, DEFAULT_SESSION_CONFIG
+from . import get_new_requests_session, urlquote_dcctx, ConcurrentUpdate, NotModified, DEFAULT_HEADERS, DEFAULT_SESSION_CONFIG
+
+class DerivaClientContext (dict):
+    """Represent Deriva-Client-Context header content.
+
+    Well-known keys (originally defined for Chaise):
+    - cid: client application ID i.e. program name
+    - wid: window ID i.e. request stream ID
+    - pid: page ID i.e. sub-stream ID
+    - action: UX action embodied by request(s)
+    - table: table upon which application is focused (if any)
+    - uinit: True if request initiated by user action
+
+    Default values to use process-wide:
+    - cid: os.path.basename(sys.argv[0]) if available
+    - wid: a random UUID
+
+    The process-wide defaults MAY be customized by mutating
+    DerivaClientContext.defaults prior to constructing instances.
+
+    """
+    defaults = {
+        'cid': os.path.basename(sys.argv[0]) if len(sys.argv) > 0 and sys.argv[0] else None,
+        'wid': uuid.uuid4().hex,
+    }
+
+    def __init__(self, *args, **kwargs):
+        """Initialize DerivaClientContext from keywords or dict-like.
+
+           Class-wide defaults are used for fields not set explicitly during construction.
+        """
+        super(DerivaClientContext, self).__init__(*args, **kwargs)
+        self.set_defaults()
+        self.prune()
+
+    def set_defaults(self, defaults=None):
+        """Set default key-values in self if key not already set."""
+        if defaults is None:
+            defaults = self.defaults
+        for k, v in defaults.items():
+            if v is not None and (k not in self or self[k] is None):
+                self[k] = v
 
+    def prune(self):
+        """Prune redundant keys to shorten context representation.
+
+           Keys with None value or uinit=False are unnecessary as these
+           are implicitly assumed when absent.
+
+        """
+        for k, v in self.items():
+            if v is None:
+                del self[k]
+        if 'uinit' in self and not self['uinit']:
+            del self['uinit']
+
+    def encoded(self):
+        """Encode self as string suitable for Deriva-Client-Context HTTP header."""
+        self.set_defaults()
+        self.prune()
+        x = urlquote_dcctx(json.dumps(self, indent=None, separators=(',', ':')))
+        print('Dcctx: %s' % x)
```

Comment:
Is this `print` statement intentional or just left over from development testing? I notice this module isn't configured to use `logging`, maybe we should add it and change this to a debug logging statement?